### ISSUE TITLE
Feature Request: Efficient rolling with strides

### ISSUE THREAD

**Issue by niowniow** [pydata/xarray#3608]
Xarray is facing the same issues in its current `rolling` implementation (`DataArrayRolling` and `DatasetRolling`) as described in [this pandas issue](https://github.com/pandas-dev/pandas/issues/15354). Namely, the `construct` methods stride parameter is applied after the rolling is computed. Technically, we are computing more than we would need to because we are throwing it away by striding.

In PR #3607 the issue is solved for the `...Rolling`'s  `__iter__` function but not for the `construct`, `reduce` and `_bottleneck_reduce` methods.
Since the way Xarray's rolling is implemented does not relies on numpy, we could introduce a sliding window function as described [here](https://github.com/numpy/numpy/issues/7753#).

Any opinions?

**Comment by niowniow** [comment]
> Previous enhancement requests asking for a `stride` argument to `rolling`: https://github.com/pandas-dev/pandas/issues/15354, https://github.com/pandas-dev/pandas/issues/22976, https://github.com/pandas-dev/pandas/issues/27654#issue-474416717, https://github.com/dask/dask/issues/4659, https://github.com/numpy/numpy/issues/7753

_Originally posted by @pilkibun in https://github.com/pandas-dev/pandas/issues/26959#issuecomment-511233955_

**Comment by GenevieveBuckley** [comment]
Is this useful/relevant for you? https://github.com/dask/dask/pull/7234

**Comment by keewis** [comment]
@dcherian, should this have been closed by #4977?

**Comment by dcherian** [comment]
No. but this should be really easy to fix.

`construct` already supports `stride`. 

`reduce` can easily support `stride` by passing it on here:
https://github.com/pydata/xarray/blob/57a4479fcd3ebc579cf00e0d6bf85007eda44b56/xarray/core/rolling.py#L441-L443

We should also add it to `_reduce_method`

bottleneck does not support stride so we can only use a `.isel` call at the end of `_bottleneck_reduce`
https://github.com/pydata/xarray/blob/57a4479fcd3ebc579cf00e0d6bf85007eda44b56/xarray/core/rolling.py#L518

Note: `sliding_window_view` does not support `stride` because it's easy to stride after constructing that view (i saw this on some numpy issue)

**Comment by niowniow** [comment]
Quickly glancing over `sliding_window_view` I didn't immediately understand how to use it with stride. Would I need to 
1. transform the DataArray to dask array using chunk (which may involve an overhead!?), 
2. then use rolling which itself uses `sliding_window_view` because its a dask array!? 
3. Then use `isel` with stride on the new dimension?

> `reduce` can easily support `stride` by passing it on here:

I think that's what I did in #3607. It's been a while

**Comment by dcherian** [comment]
`sliding_window_view` is a numpy function (see `npcompat.py`), so you need not transform to dask.

For `reduce` I think we just have to pass `stride` to `_construct` as in https://github.com/pydata/xarray/issues/3608#issuecomment-811156323 and for `bottleneck` insert the `.isel` call (copied from the end of `_construct`) after the DataArray is constructed.

**Comment by kmsquire** [comment]
Question: instead of adding `stride` to `reduce` and `_reduce_method`, why not add it as a member of `DataArrayRolling` directly?  This would allow, e.g., `__iter__` to use it as well, and seems like a cleaner interface.  

I've been confused why some parameters are available only in `construct` (`stride`, `fill_value`), some are available both in `construct` and in the `DataArrayRolling` constructor (`keep_attrs`), and some are only available in the constructor (`min_periods`, `center`, and soon `pad`).

**Comment by niowniow** [comment]
You are right. It's quite confusing. I've already added a `stride` parameter in my PR #3607 
I didn't follow through with it and at the moment the checks are not successful anymore.
Maybe someone else could give an opinion on the pro/cons of a `stride` parameter in rolling?
